### ISSUE TITLE
Data Loaders in Nx?

### ISSUE THREAD

**Issue by krstopro** [elixir-nx/nx#1468]
From what I saw so far, the conventional way to iterate the data in batches is to use functions from `Enum` and `Stream`. Are there already abstractions for doing this in Nx, e.g. like [DataLoader](https://pytorch.org/docs/stable/data.html#torch.utils.data.DataLoader) in PyTorch?

**Comment by krstopro** [comment]
I think I saw the same patterns being used several times. Basically, data shuffling using `Enum.shuffle/1`, batching with `Nx.to_batched/3` or `Stream.chunk_every/4`, augmentation with `Stream.map/2`, etc. Something like [data_stream in this notebook](https://github.com/elixir-nx/axon/blob/main/notebooks/vision/horses_or_humans.livemd).

Feel free to close the issue if you feel like it's impossible to wrap this in a module.

**Comment by josevalim** [comment]
I think the biggest question is: should we wrap it? Why not use functions that compose neatly? Isn't that better than a large interface with several options that hide how they relate to each other?

**Comment by krstopro** [comment]
I am not sure myself, which is another reason I asked. There aren't that many options that a loader should support: batch size, whether to shuffle the dataset, number of workers, whether to drop the incomplete batch, collate function, etc. Perhaps releasing a notebook with a conventional way to do it might be a better idea.

**Comment by josevalim** [comment]
Yeah, Axon already has a guide and maybe we can have one in Scholar that starts with a dataset and manipulates it, but for now I don't think it is a Nx specific concern. :) Thanks!

**Comment by krstopro** [comment]
Agreed!
